llvm/include/llvm/IR/VectorUtils.h

```diff
@@ -0,0 +1,53 @@
+//===----------- VectorUtils.h -  Vector type utility functions -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/DerivedTypes.h"
+
+namespace llvm {
+
+/// A helper function for converting Scalar types to vector types. If
+/// the incoming type is void, we return void. If the EC represents a
+/// scalar, we return the scalar type.
+inline Type *ToVectorTy(Type *Scalar, ElementCount EC) {
+  if (Scalar->isVoidTy() || Scalar->isMetadataTy() || EC.isScalar())
+    return Scalar;
+  return VectorType::get(Scalar, EC);
+}
+
+inline Type *ToVectorTy(Type *Scalar, unsigned VF) {
+  return ToVectorTy(Scalar, ElementCount::getFixed(VF));
+}
+
+/// A helper for converting to wider (vector) types. For scalar types, this is
+/// equivalent to calling `ToVectorTy`. For struct types, this returns a new
+/// struct where each element type has been widened to a vector type. Note: Only
+/// unpacked literal struct types are supported.
+Type *ToWideTy(Type *Ty, ElementCount EC);
```

Comment:
> Should the result type be a VectorType ?

The result type can be a vector, or a wide struct literal (e.g. `{ <4 x float>, <4 x float> }`). "Wide" here means that all elements of the struct literal are vectors of the same element count.

Reply:
Of course, not sure how I didn't realise that when I wrote the comment :)

Reply:
> What about creating a single interface called widenScalarToVector that applies to both scalar types and aggregates (with scalar elements)? That way you can remove the distinction between ToVectorTy and ToWideTy.

I thought it made sense to keep the distinction between "wide" types and vector types. Many code paths only deal with vector types, and I thought it may be confusing to refer to wide structs as vectors.